File: services/domain-processor-v2/test-all-api-keys.js
#!/usr/bin/env node

// TEST ALL API KEYS - Validate each provider with correct models and endpoints
const axios = require('axios');

const PROVIDERS = [
  {
    name: 'OpenAI',
    apiKey: process.env.OPENAI_API_KEY,
    test: async (apiKey) => {
      const response = await axios.post('https://api.openai.com/v1/chat/completions', {
        model: 'gpt-4o-mini',
        messages: [{ role: 'user', content: 'Say "test ok" in 2 words' }],
        max_tokens: 10
      }, {
        headers: { 'Authorization': `Bearer ${apiKey}` }
      });
      return response.data.choices[0].message.content;
    }
  },
  
  {
    name: 'Anthropic',
    apiKey: process.env.ANTHROPIC_API_KEY,
    test: async (apiKey) => {
      const response = await axios.post('https://api.anthropic.com/v1/messages', {
        model: 'claude-3-haiku-20240307',
        messages: [{ role: 'user', content: 'Say "test ok" in 2 words' }],
        max_tokens: 10
      }, {
        headers: { 
          'x-api-key': apiKey,
          'anthropic-version': '2023-06-01',
          'content-type': 'application/json'
        }
      });
      return response.data.content[0].text;
    }
  },
  
  {
    name: 'DeepSeek',
    apiKey: process.env.DEEPSEEK_API_KEY,
    test: async (apiKey) => {
      const response = await axios.post('https://api.deepseek.com/v1/chat/completions', {
        model: 'deepseek-chat',
        messages: [{ role: 'user', content: 'Say "test ok" in 2 words' }],
        max_tokens: 10
      }, {
        headers: { 'Authorization': `Bearer ${apiKey}` }
      });
      return response.data.choices[0].message.content;
    }
  },
  
  {
    name: 'Mistral',
    apiKey: process.env.MISTRAL_API_KEY,
    test: async (apiKey) => {
      const response = await axios.post('https://api.mistral.ai/v1/chat/completions', {
        model: 'mistral-small-latest',
        messages: [{ role: 'user', content: 'Say "test ok" in 2 words' }],
        max_tokens: 10
      }, {
        headers: { 
          'Authorization': `Bearer ${apiKey}`,
          'Content-Type': 'application/json'
        }
      });
      return response.data.choices[0].message.content;
    }
  },
  
  {
    name: 'Cohere',
    apiKey: process.env.COHERE_API_KEY,
    test: async (apiKey) => {
      const response = await axios.post('https://api.cohere.ai/v1/chat', {
        model: 'command-r',
        message: 'Say "test ok" in 2 words',
        max_tokens: 10
      }, {
        headers: { 
          'Authorization': `Bearer ${apiKey}`,
          'Content-Type': 'application/json'
        }
      });
      return response.data.text;
    }
  },
  
  {
    name: 'Together',
    apiKey: process.env.TOGETHER_API_KEY,
    test: async (apiKey) => {
      const response = await axios.post('https://api.together.xyz/v1/chat/completions', {
        model: 'meta-llama/Llama-3-8b-chat-hf',
        messages: [{ role: 'user', content: 'Say "test ok" in 2 words' }],
        max_tokens: 10
      }, {
        headers: { 'Authorization': `Bearer ${apiKey}` }
      });
      return response.data.choices[0].message.content;
    }
  },
  
  {
    name: 'Groq',
    apiKey: process.env.GROQ_API_KEY,
    test: async (apiKey) => {
      // Note: Groq model names changed recently
      const models = [
        'llama-3.2-90b-text-preview',
        'llama3-groq-70b-8192-tool-use-preview',
        'mixtral-8x7b-32768',
        'llama3-8b-8192'
      ];
      
      for (const model of models) {
        try {
          const response = await axios.post('https://api.groq.com/openai/v1/chat/completions', {
            model: model,
            messages: [{ role: 'user', content: 'Say "test ok" in 2 words' }],
            max_tokens: 10
          }, {
            headers: { 'Authorization': `Bearer ${apiKey}` }
          });
          console.log(`   → Working model: ${model}`);
          return response.data.choices[0].message.content;
        } catch (e) {
          continue;
        }
      }
      throw new Error('No working Groq models found');
    }
  },
  
  {
    name: 'XAI (Grok)',
    apiKey: process.env.XAI_API_KEY,
    test: async (apiKey) => {
      // XAI uses OpenAI-compatible endpoint
      const response = await axios.post('https://api.x.ai/v1/chat/completions', {
        model: 'grok-beta',
        messages: [{ role: 'user', content: 'Say "test ok" in 2 words' }],
        max_tokens: 10
      }, {
        headers: { 'Authorization': `Bearer ${apiKey}` }
      });
      return response.data.choices[0].message.content;
    }
  },
  
  {
    name: 'Google (Gemini)',
    apiKey: process.env.GOOGLE_API_KEY,
    test: async (apiKey) => {
      // Google uses different endpoint structure
      const response = await axios.post(
        `https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key=${apiKey}`,
        {
          contents: [{
            parts: [{
              text: 'Say "test ok" in 2 words'
            }]
          }]
        }
      );
      return response.data.candidates[0].content.parts[0].text;
    }
  },
  
  {
    name: 'AI21',
    apiKey: process.env.AI21_API_KEY,
    test: async (apiKey) => {
      const response = await axios.post('https://api.ai21.com/studio/v1/j2-ultra/complete', {
        prompt: 'Say "test ok" in 2 words:',
        maxTokens: 10,
        temperature: 0.7
      }, {
        headers: { 
          'Authorization': `Bearer ${apiKey}`,
          'Content-Type': 'application/json'
        }
      });
      return response.data.completions[0].data.text;
    }
  },
  
  {
    name: 'OpenRouter',
    apiKey: process.env.OPENROUTER_API_KEY,
    test: async (apiKey) => {
      const response = await axios.post('https://openrouter.ai/api/v1/chat/completions', {
        model: 'meta-llama/llama-3.1-70b-instruct',
        messages: [{ role: 'user', content: 'Say "test ok" in 2 words' }]
      }, {
        headers: { 
          'Authorization': `Bearer ${apiKey}`,
          'HTTP-Referer': 'https://llmpagerank.com',
          'Content-Type': 'application/json'
        }
      });
      return response.data.choices[0].message.content;
    }
  },
  
  {
    name: 'Perplexity',
    apiKey: process.env.PERPLEXITY_API_KEY,
    test: async (apiKey) => {
      const models = [
        'llama-3.1-sonar-small-128k-online',
        'llama-3.1-sonar-large-128k-online',
        'llama-3.1-sonar-huge-128k-online'
      ];
      
      for (const model of models) {
        try {
          const response = await axios.post('https://api.perplexity.ai/chat/completions', {
            model: model,
            messages: [{ role: 'user', content: 'Say "test ok" in 2 words' }]
          }, {
            headers: { 
              'Authorization': `Bearer ${apiKey}`,
              'Content-Type': 'application/json'
            }
          });
          console.log(`   → Working model: ${model}`);
          return response.data.choices[0].message.content;
        } catch (e) {
          continue;
        }
      }
      throw new Error('No working Perplexity models found');
    }
  }
];

async function testAllProviders() {
  console.log('🧪 TESTING ALL API KEYS AND MODELS');
  console.log('=====================================\n');
  
  const results = {
    working: [],
    failed: []
  };
  
  for (const provider of PROVIDERS) {
    process.stdout.write(`Testing ${provider.name}... `);
    
    if (!provider.apiKey) {
      console.log('❌ No API key');
      results.failed.push({ name: provider.name, error: 'No API key' });
      continue;
    }
    
    try {
      const result = await provider.test(provider.apiKey);
      console.log(`✅ Working! Response: "${result}"`);
      results.working.push(provider.name);
    } catch (error) {
      const errorMsg = error.response?.data?.error?.message || 
                      error.response?.data?.message || 
                      error.response?.status || 
                      error.message;
      console.log(`❌ Failed: ${errorMsg}`);
      
      // Show detailed error for debugging
      if (error.response?.data) {
        console.log(`   Details: ${JSON.stringify(error.response.data).substring(0, 200)}`);
      }
      
      results.failed.push({ 
        name: provider.name, 
        error: errorMsg,
        status: error.response?.status
      });
    }
    
    // Small delay to avoid rate limits
    await new Promise(resolve => setTimeout(resolve, 500));
  }
  
  console.log('\n=====================================');
  console.log('📊 SUMMARY');
  console.log('=====================================');
  console.log(`✅ Working: ${results.working.length} providers`);
  results.working.forEach(p => console.log(`   - ${p}`));
  
  console.log(`\n❌ Failed: ${results.failed.length} providers`);
  results.failed.forEach(p => console.log(`   - ${p.name}: ${p.error} ${p.status ? `(HTTP ${p.status})` : ''}`));
  
  console.log('\n💡 RECOMMENDATIONS:');
  results.failed.forEach(p => {
    if (p.status === 401) {
      console.log(`   ${p.name}: Check API key format and validity`);
    } else if (p.status === 400) {
      console.log(`   ${p.name}: Model name or request format may be incorrect`);
    } else if (p.status === 404) {
      console.log(`   ${p.name}: API endpoint URL may be wrong`);
    }
  });
}

testAllProviders().catch(console.error);

File: test_all_api_keys.py
#!/usr/bin/env python3
"""Test ALL API keys including secondary ones"""
import requests
import json

print("🔍 TESTING ALL API KEYS (PRIMARY & SECONDARY)")
print("=" * 50)

# All API keys from render.yaml
api_configs = {
    'OPENAI': [
        ('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', 'OPENAI_API_KEY'),
        # Need to find OPENAI_API_KEY_2, 3, 4
    ],
    'ANTHROPIC': [
        ('sk-ant-api03-jZa-W0Cyk3Z_s7vF_dLYJkP2YYiclqS0d8M-dO15s_j4fPFnNu_kFPXnCx3aK-pD-O8D3_DVqFMZ0rBJJ6Kg5g-x2nA8AAA', 'ANTHROPIC_API_KEY'),
        # Need ANTHROPIC_API_KEY_2
    ],
    'DEEPSEEK': [
        ('sk-a03c67f1fdd74c139faa0ad69b44a0fa', 'DEEPSEEK_API_KEY'),
        # Need DEEPSEEK_API_KEY_2, 3
    ],
    'MISTRAL': [
        ('ft2Xg7JfRU7OXoBQnrmIlLQdVJQQO89Z', 'MISTRAL_API_KEY'),
        # Need MISTRAL_API_KEY_2
    ],
    'XAI': [
        ('xai-TvMNjOdmQG8wFYI8nplKvopQlflnCSDo1fwmUl7XzQ9TLXrGZcJ4OJnJXGRRn7pjP7VKJBHQAyU4Yonc', 'XAI_API_KEY'),
        # Need XAI_API_KEY_2
    ],
    'TOGETHER': [
        ('9e3ba0c46dd44a97d19bb02c86bc79fdbbbe4acdad62c3c088c96cc08758c8f4', 'TOGETHER_API_KEY'),
        # Need TOGETHER_API_KEY_2, 3
    ],
    'PERPLEXITY': [
        ('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', 'PERPLEXITY_API_KEY'),
        # Need PERPLEXITY_API_KEY_2
    ],
    'GOOGLE': [
        ('AIzaSyDi-i8I9BiL7E36skCmR6BQXNO7Y5LHnxg', 'GOOGLE_API_KEY'),
        # Need GOOGLE_API_KEY_2
    ]
}

def test_openai(key):
    try:
        resp = requests.post(
            'https://api.openai.com/v1/chat/completions',
            headers={'Authorization': f'Bearer {key}', 'Content-Type': 'application/json'},
            json={'model': 'gpt-3.5-turbo', 'messages': [{'role': 'user', 'content': 'test'}], 'max_tokens': 5},
            timeout=5
        )
        return resp.status_code == 200
    except:
        return False

def test_anthropic(key):
    try:
        resp = requests.post(
            'https://api.anthropic.com/v1/messages',
            headers={'x-api-key': key, 'anthropic-version': '2023-06-01', 'Content-Type': 'application/json'},
            json={'model': 'claude-3-haiku-20240307', 'messages': [{'role': 'user', 'content': 'test'}], 'max_tokens': 5},
            timeout=5
        )
        return resp.status_code == 200
    except:
        return False

def test_deepseek(key):
    try:
        resp = requests.post(
            'https://api.deepseek.com/v1/chat/completions',
            headers={'Authorization': f'Bearer {key}', 'Content-Type': 'application/json'},
            json={'model': 'deepseek-chat', 'messages': [{'role': 'user', 'content': 'test'}], 'max_tokens': 5},
            timeout=5
        )
        return resp.status_code == 200
    except:
        return False

def test_mistral(key):
    try:
        resp = requests.post(
            'https://api.mistral.ai/v1/chat/completions',
            headers={'Authorization': f'Bearer {key}', 'Content-Type': 'application/json'},
            json={'model': 'mistral-tiny', 'messages': [{'role': 'user', 'content': 'test'}], 'max_tokens': 5},
            timeout=5
        )
        return resp.status_code == 200
    except:
        return False

def test_xai(key):
    try:
        resp = requests.post(
            'https://api.x.ai/v1/chat/completions',
            headers={'Authorization': f'Bearer {key}', 'Content-Type': 'application/json'},
            json={'model': 'grok-beta', 'messages': [{'role': 'user', 'content': 'test'}], 'max_tokens': 5},
            timeout=5
        )
        return resp.status_code == 200
    except:
        return False

def test_together(key):
    try:
        resp = requests.post(
            'https://api.together.xyz/v1/chat/completions',
            headers={'Authorization': f'Bearer {key}', 'Content-Type': 'application/json'},
            json={'model': 'mistralai/Mistral-7B-Instruct-v0.1', 'messages': [{'role': 'user', 'content': 'test'}], 'max_tokens': 5},
            timeout=5
        )
        return resp.status_code == 200
    except:
        return False

def test_perplexity(key):
    try:
        resp = requests.post(
            'https://api.perplexity.ai/chat/completions',
            headers={'Authorization': f'Bearer {key}', 'Content-Type': 'application/json'},
            json={'model': 'mistral-7b-instruct', 'messages': [{'role': 'user', 'content': 'test'}], 'max_tokens': 5},
            timeout=5
        )
        return resp.status_code == 200
    except:
        return False

def test_google(key):
    try:
        resp = requests.post(
            f'https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key={key}',
            headers={'Content-Type': 'application/json'},
            json={'contents': [{'parts': [{'text': 'test'}]}]},
            timeout=5
        )
        return resp.status_code == 200
    except:
        return False

# Test functions map
test_functions = {
    'OPENAI': test_openai,
    'ANTHROPIC': test_anthropic,
    'DEEPSEEK': test_deepseek,
    'MISTRAL': test_mistral,
    'XAI': test_xai,
    'TOGETHER': test_together,
    'PERPLEXITY': test_perplexity,
    'GOOGLE': test_google
}

# Test all keys
working_providers = []
for provider, keys in api_configs.items():
    print(f"\n{provider}:")
    test_func = test_functions[provider]
    
    for i, (key, env_name) in enumerate(keys):
        works = test_func(key)
        status = "✅ Working" if works else "❌ Failed"
        print(f"  Key {i+1} ({env_name}): {status}")
        if works:
            working_providers.append(provider)
            break  # One working key is enough

print(f"\n📊 SUMMARY: {len(working_providers)}/8 providers have at least one working key")
print(f"✅ Working: {', '.join(working_providers)}")

# Check against what's actually being used
print("\n🔍 Checking against current usage...")
import psycopg2
conn = psycopg2.connect('postgresql://raw_capture_db_user:wjFesUM8ISNEvE2b4kZtRAKgGYJVtKK5@dpg-d11fqgndiees73fb35dg-a.oregon-postgres.render.com/raw_capture_db')
cursor = conn.cursor()
cursor.execute("""
    SELECT DISTINCT model FROM domain_responses 
    WHERE created_at > NOW() - INTERVAL '2 hours'
    ORDER BY model
""")
active_models = [row[0] for row in cursor.fetchall()]
print(f"Currently active in DB: {', '.join(active_models)}")

print("\n💡 INSIGHT:")
print("If Render has the secondary API keys that we don't have here,")
print("those might be the ones actually working in production!")